binding.cc

```diff
@@ -2194,608 +2288,663 @@ sn_crypto_secretstream_xchacha20poly1305_pull(
+  crypto_secretstream_xchacha20poly1305_rekey(state_data);
 }
 
-typedef struct sn_async_task_t {
+struct sn_async_task_t {
   uv_work_t task;
-
-  enum {
-    sn_async_task_promise,
-    sn_async_task_callback
-  } type;
-
-  void *req;
+  js_env_t *env;
+  js_persistent_t<js_function_t<void, int>> cb;
   int code;
+};
+
+struct sn_async_pwhash_request : sn_async_task_t {
+  js_persistent_t<js_arraybuffer_t> out_ref;
+  std::span<uint8_t> out;
+
+  js_persistent_t<js_arraybuffer_t> pwd_ref;
+  std::span<char> pwd;
+
+  js_persistent_t<js_arraybuffer_t> salt_ref;
+  std::span<uint8_t> salt;
+
+  uint64_t opslimit;
+  size_t memlimit;
+  int alg;
+};
+
+static void
+async_pwhash_execute(uv_work_t *uv_req) {
+  auto req = reinterpret_cast<sn_async_pwhash_request *>(uv_req);
+  req->code = crypto_pwhash(
+    req->out.data(),
+    req->out.size(),
+    req->pwd.data(),
+    req->pwd.size(),
+    req->salt.data(),
+    req->opslimit,
+    req->memlimit,
+    req->alg
+  );
+}
 
-  js_deferred_t *deferred;
-  js_ref_t *cb;
-} sn_async_task_t;
-
-typedef struct sn_async_pwhash_request {
-  js_env_t *env;
-  js_ref_t *out_ref;
-  unsigned char *out_data;
-  size_t out_size;
-  js_ref_t *pwd_ref;
-  const char *pwd_data;
-  size_t pwd_size;
-  js_ref_t *salt_ref;
-  unsigned char *salt;
-  uint32_t opslimit;
-  uint32_t memlimit;
-  uint32_t alg;
-} sn_async_pwhash_request;
-
-static void async_pwhash_execute (uv_work_t *uv_req) {
-  sn_async_task_t *task = (sn_async_task_t *) uv_req;
-  sn_async_pwhash_request *req = (sn_async_pwhash_request *) task->req;
-  task->code = crypto_pwhash(req->out_data,
-                             req->out_size,
-                             req->pwd_data,
-                             req->pwd_size,
-                             req->salt,
-                             req->opslimit,
-                             req->memlimit,
-                             req->alg);
-}
-
-static void async_pwhash_complete (uv_work_t *uv_req, int status) {
+static void
+async_pwhash_complete(uv_work_t *uv_req, int status) {
   int err;
-  sn_async_task_t *task = (sn_async_task_t *) uv_req;
-  sn_async_pwhash_request *req = (sn_async_pwhash_request *) task->req;
+
+  auto req = reinterpret_cast<sn_async_pwhash_request *>(uv_req);
 
   js_handle_scope_t *scope;
   err = js_open_handle_scope(req->env, &scope);
   assert(err == 0);
 
-  js_value_t *global;
-  err = js_get_global(req->env, &global);
+  js_function_t<void, int> callback;
+  err = js_get_reference_value(req->env, req->cb, callback);
   assert(err == 0);
 
-  SN_ASYNC_COMPLETE("failed to compute password hash")
+  err = js_call_function_with_checkpoint(req->env, callback, req->code);
+  assert(err != js_pending_exception);
 
   err = js_close_handle_scope(req->env, scope);
   assert(err == 0);
 
-  err = js_delete_reference(req->env, req->out_ref);
+  delete req;
+}
+
+static inline void
+sn_crypto_pwhash_async(
+  js_env_t *env,
+  js_receiver_t,
+
+  js_arraybuffer_t out,
```

Comment:
Can use `js_arraybuffer_span_t`.

Reply:
cannot pass `js_typedarray_span_t` to  `js_create_reference()`:
```
error: cannot cast from type 'const js_arraybuffer_span_t' to pointer type 'js_value_t *' 
```

Reply:
Very good point, `js_typedarray_span_t` doesn't maintain a reference to a possible underlying ABI handle.